osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs

```diff
@@ -282,13 +296,78 @@ public override Task ChangeUserMods(IEnumerable<APIMod> newMods)
             return Task.CompletedTask;
         }
 
+        private CancellationTokenSource? countdownFinishSource;
+        private CancellationTokenSource? countdownStopSource;
+        private Task countdownTask = Task.CompletedTask;
+
+        public void FinishCountdown() => countdownFinishSource?.Cancel();
+
         public override async Task SendMatchRequest(MatchUserRequest request)
         {
             Debug.Assert(Room != null);
             Debug.Assert(LocalUser != null);
 
             switch (request)
             {
+                case StartMatchCountdownRequest matchCountdownRequest:
+                    countdownStopSource?.Cancel();
+
+                    var stopSource = countdownStopSource = new CancellationTokenSource();
+                    var finishSource = countdownFinishSource = new CancellationTokenSource();
+                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, finishSource.Token);
+                    var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
+
+                    Task lastCountdownTask = countdownTask;
+                    countdownTask = start();
+
+                    async Task start()
+                    {
+                        await lastCountdownTask;
+
+                        Schedule(() =>
+                        {
+                            if (stopSource.IsCancellationRequested)
+                                return;
+
+                            Room.Countdown = countdown;
+                            MatchEvent(new CountdownChangedEvent { Countdown = countdown });
+                        });
+
+                        try
+                        {
+                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        Schedule(() =>
+                        {
+                            if (Room.Countdown != countdown)
+                                return;
+
+                            Room.Countdown = null;
+                            MatchEvent(new CountdownChangedEvent { Countdown = null });
+
+                            using (cancellationSource)
```

Comment:
I'm very confused by this, and the rest of the cancellation logic.

Reply:
I'm confused why you're confused.

Reply:
I'll try to explain better:

I've never seen us use a `using` statement like this elsewhere, and I'm confused why it's done in this way. Is the overall goal to dispose of this source now that we're done with it? If so, is there any reason not to call `cancellationSource.Dispose()`?

Reply:
Yes, it is purely to dispose it. That being said, this code differs from the server version in that it has a precheck to make sure the countdown matches (due to the schedules), and it also needs to be disposed in that case.

I'll change it to a manual dispose.

Reply:
Actually, I disagree with a manual dispose. I will keep using a `using` because I want it to dispose after either of the 3 potential `return`s have been taken.

Reply:
Can the `using` be moved out to a higher level then?

```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
index 1f20292437..8cd62c868a 100644
--- a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
@@ -314,7 +314,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     var stopSource = countdownStopSource = new CancellationTokenSource();
                     var finishSource = countdownFinishSource = new CancellationTokenSource();
-                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, finishSource.Token);
                     var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
 
                     Task lastCountdownTask = countdownTask;
@@ -322,41 +321,44 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     async Task start()
                     {
-                        await lastCountdownTask;
-
-                        Schedule(() =>
+                        using (var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, finishSource.Token))
                         {
-                            if (stopSource.IsCancellationRequested)
-                                return;
+                            await lastCountdownTask;
 
-                            Room.Countdown = countdown;
-                            MatchEvent(new CountdownChangedEvent { Countdown = countdown });
-                        });
+                            Schedule(() =>
+                            {
+                                if (stopSource.IsCancellationRequested)
+                                    return;
 
-                        try
-                        {
-                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
-                        }
-                        catch (OperationCanceledException)
-                        {
-                        }
+                                Room.Countdown = countdown;
+                                MatchEvent(new CountdownChangedEvent { Countdown = countdown });
+                            });
 
-                        Schedule(() =>
-                        {
-                            if (Room.Countdown != countdown)
-                                return;
-
-                            Room.Countdown = null;
-                            MatchEvent(new CountdownChangedEvent { Countdown = null });
+                            try
+                            {
+                                await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
+                            }
+                            catch (OperationCanceledException)
+                            {
+                            }
 
-                            using (cancellationSource)
+                            Schedule(() =>
                             {
-                                if (stopSource.Token.IsCancellationRequested)
+                                if (Room.Countdown != countdown)
                                     return;
-                            }
 
-                            StartMatch().WaitSafely();
-                        });
+                                Room.Countdown = null;
+                                MatchEvent(new CountdownChangedEvent { Countdown = null });
+
+                                using (cancellationSource)
+                                {
+                                    if (stopSource.Token.IsCancellationRequested)
+                                        return;
+                                }
+
+                                StartMatch().WaitSafely();
+                            });
+                        }
                     }
 
                     break;
```

Reply:
No, because the coed inside is scheduling and asyncing...

Reply:
Is there some special meaning behind the point of disposal? Does it trigger some behaviour I'm unaware of?

Reply:
Let's simplify it.

```csharp
cancellationSource = ...;

try
{
    Schedule(() => doSomethingWithCancellationSource());
}
finally
{
    cancellationSource.Dispose();
}
```
This is equivalent to `using (cancellationSource)`. Do you see the issue now? The source will be disposed _before_ `doSomethingWithCancellationSource()` is run, because it's scheduled, which means that code will throw an exception.

The current code is a little bit different to the above case, specifically:
1. It doesn't use the disposed cancellation source in the schedules.
2. When referring to the cancellation sources in the schedules, it always refers to the ones that _aren't_ disposed.

But those differences are not enough for me to justify changing this code as per your proposal, because even just the possibility of using the disposed cancellation source in the schedules makes me scared.

Reply:
I have to say that I thought this explanation made sense, but on checking a second time... the linked source is only used in the `Schedule` for the purpose of disposing it. Unless I'm missing something I think my proposal is still correct.

```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
index ea991af914..3a59495269 100644
--- a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
@@ -318,7 +318,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     var stopSource = countdownStopSource = new CancellationTokenSource();
                     var skipSource = countdownSkipSource = new CancellationTokenSource();
-                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token);
                     var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
 
                     Task lastCountdownTask = countdownTask;
@@ -326,29 +325,29 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     async Task start()
                     {
-                        await lastCountdownTask;
-
-                        Schedule(() =>
+                        using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token))
                         {
-                            if (stopSource.IsCancellationRequested)
-                                return;
+                            await lastCountdownTask;
 
-                            Room.Countdown = countdown;
-                            MatchEvent(new CountdownChangedEvent { Countdown = countdown });
-                        });
+                            Schedule(() =>
+                            {
+                                if (stopSource.IsCancellationRequested)
+                                    return;
 
-                        try
-                        {
-                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
-                        }
-                        catch (OperationCanceledException)
-                        {
-                            // Clients need to be notified of cancellations in the following code.
-                        }
+                                Room.Countdown = countdown;
+                                MatchEvent(new CountdownChangedEvent { Countdown = countdown });
+                            });
 
-                        Schedule(() =>
-                        {
-                            using (cancellationSource)
+                            try
+                            {
+                                await Task.Delay(matchCountdownRequest.Duration, linkedSource.Token).ConfigureAwait(false);
+                            }
+                            catch (OperationCanceledException)
+                            {
+                                // Clients need to be notified of cancellations in the following code.
+                            }
+
+                            Schedule(() =>
                             {
                                 if (Room.Countdown != countdown)
                                     return;
@@ -356,12 +355,10 @@ async Task start()
                                 Room.Countdown = null;
                                 MatchEvent(new CountdownChangedEvent { Countdown = null });
 
-                                if (stopSource.Token.IsCancellationRequested)
-                                    return;
-
-                                StartMatch().WaitSafely();
-                            }
-                        });
+                                if (!stopSource.Token.IsCancellationRequested)
+                                    StartMatch().WaitSafely();
+                            });
+                        }
                     }
 
                     break;
```

diff also has slightly renaming that i feel works better.

Reply:
Slightly simpler:

```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
index ea991af914..f482f3826c 100644
--- a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
@@ -318,7 +318,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     var stopSource = countdownStopSource = new CancellationTokenSource();
                     var skipSource = countdownSkipSource = new CancellationTokenSource();
-                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token);
                     var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
 
                     Task lastCountdownTask = countdownTask;
@@ -337,18 +336,18 @@ async Task start()
                             MatchEvent(new CountdownChangedEvent { Countdown = countdown });
                         });
 
-                        try
+                        using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token))
                         {
-                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
-                        }
-                        catch (OperationCanceledException)
-                        {
-                            // Clients need to be notified of cancellations in the following code.
-                        }
+                            try
+                            {
+                                await Task.Delay(matchCountdownRequest.Duration, linkedSource.Token).ConfigureAwait(false);
+                            }
+                            catch (OperationCanceledException)
+                            {
+                                // Clients need to be notified of cancellations in the following code.
+                            }
 
-                        Schedule(() =>
-                        {
-                            using (cancellationSource)
+                            Schedule(() =>
                             {
                                 if (Room.Countdown != countdown)
                                     return;
@@ -356,12 +355,10 @@ async Task start()
                                 Room.Countdown = null;
                                 MatchEvent(new CountdownChangedEvent { Countdown = null });
 
-                                if (stopSource.Token.IsCancellationRequested)
-                                    return;
-
-                                StartMatch().WaitSafely();
-                            }
-                        });
+                                if (!stopSource.Token.IsCancellationRequested)
+                                    StartMatch().WaitSafely();
+                            });
+                        }
                     }
 
                     break;
```

posting separate rather than edit to make sure it's not missed.

Reply:
In fact...

```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
index ea991af914..85c4bdc333 100644
--- a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
@@ -318,7 +318,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)
 
                     var stopSource = countdownStopSource = new CancellationTokenSource();
                     var skipSource = countdownSkipSource = new CancellationTokenSource();
-                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token);
                     var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
 
                     Task lastCountdownTask = countdownTask;
@@ -339,7 +338,8 @@ async Task start()
 
                         try
                         {
-                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
+                            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token))
+                                await Task.Delay(matchCountdownRequest.Duration, linkedSource.Token).ConfigureAwait(false);
                         }
                         catch (OperationCanceledException)
                         {
@@ -348,19 +348,14 @@ async Task start()
 
                         Schedule(() =>
                         {
-                            using (cancellationSource)
-                            {
-                                if (Room.Countdown != countdown)
-                                    return;
-
-                                Room.Countdown = null;
-                                MatchEvent(new CountdownChangedEvent { Countdown = null });
+                            if (Room.Countdown != countdown)
+                                return;
 
-                                if (stopSource.Token.IsCancellationRequested)
-                                    return;
+                            Room.Countdown = null;
+                            MatchEvent(new CountdownChangedEvent { Countdown = null });
 
+                            if (!stopSource.Token.IsCancellationRequested)
                                 StartMatch().WaitSafely();
-                            }
                         });
                     }
 
```

Reply:
I'd be fine with the following:
```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
index ea991af914..2ed0d2dd63 100644
--- a/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
@@ -318,7 +318,6 @@ public override async Task SendMatchRequest(MatchUserRequest request)

                     var stopSource = countdownStopSource = new CancellationTokenSource();
                     var skipSource = countdownSkipSource = new CancellationTokenSource();
-                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token);
                     var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };

                     Task lastCountdownTask = countdownTask;
@@ -339,7 +338,8 @@ async Task start()

                         try
                         {
-                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
+                            using (var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, skipSource.Token))
+                                await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
                         }
                         catch (OperationCanceledException)
                         {
@@ -348,15 +348,19 @@ async Task start()

                         Schedule(() =>
                         {
-                            using (cancellationSource)
+                            using (stopSource)
+                            using (skipSource)
                             {
                                 if (Room.Countdown != countdown)
                                     return;

+                                countdownStopSource = null;
+                                countdownSkipSource = null;
+
                                 Room.Countdown = null;
                                 MatchEvent(new CountdownChangedEvent { Countdown = null });

-                                if (stopSource.Token.IsCancellationRequested)
+                                if (stopSource.IsCancellationRequested)
                                     return;

                                 StartMatch().WaitSafely();
```

Reply:
I'm not 100% on the source null sets there, aren't they initialised outside the `Schedule` making this potentially unsafe?

Also can the `using (stopSource)` using statements be moved up to the construction calls for simplicity?

Reply:
> I'm not 100% on the source null sets there, aren't they initialised outside the `Schedule` making this potentially unsafe?

I can add a `ThreadSafety` assert. The outer code must run on the update thread for it to be thread-safe at all. The null sets themselves are required because we're disposing them now any any further calls to `.Cancel()` will throw an exception.

> Also can the `using (stopSource)` using statements be moved up to the construction calls for simplicity?

No. This circles back to the original point.

Reply:
Please add a `ThreadSafety` assert, yep.

Rather than `using` can we settle on `Dispose` calls before the ctor? I should remove the need for nulling them too (which I think reads pretty weird).

```csharp
countdownStopSource?.Dispose()
countdownStopSource = new...
```

Reply:
> Rather than `using` can we settle on `Dispose` calls before the ctor?

And we just leak cancellation sources? Just so we're clear, I agree with aggressively disposing `CancellationTokenSources` rather than letting them hang, but what you're suggesting would imply there's always a non-disposed CTS.

Reply:
Yeah, proposal would be "leaking" one max. Which I wouldn't call leaking.

Reply:
Note however, that osu-server-spectator which does the same thing, will have to dispose them as per my above code.

Reply:
I've refactored this based on the above + discord conversation. It does mean that we're leaking non-linked CTSs, however we haven't yet found that to be detrimental in other parts of code (such as the hub connector, for example).

Note that the osu-server-spectator implementation differs in this regard because I don't want to take any risks there: https://github.com/ppy/osu-server-spectator/pull/109/files#diff-18f60f089f2383925c0417daefb117e8739cb1c75fd9832f2f9d6f2cdea561cfR163-R190